apbs/input_file/calculate/finite_difference.py

```diff
@@ -0,0 +1,1899 @@
+                    f"({self._lengths}) or spacings ({self._spacings}) is None."
+                )
+        return self._counts
+
+    @counts.setter
+    def counts(self, value):
+        if len(value) != 3:
+            raise IndexError(f"Counts are {len(value)} rather than 3.")
+        self._counts = []
+        for i, elem in enumerate(value):
+            if (elem < 2 ** MIN_LEVEL) or not isinstance(elem, int):
+                self._counts = None
+                raise TypeError(
+                    f"Vector {value} does not contain positive integers."
+                )
+            self._counts.append(elem)
+        _LOGGER.debug(f"Target grid nunmbers are {self._counts}.")
+        self.adjust_counts()
+        _LOGGER.debug(
+            f"Actual grid number are {self._counts} with level {self.levels}."
+        )
+
+    @property
+    def spacings(self) -> list:
+        """Spacings of the grid in a finite-difference Poisson-Boltzmann
+        calculation.
+
+        :returns: 3-vector of positive numbers indicating grid spacings in the
+          x-, y-, and z-directions in Å. The spacings may be different in each
+          direction.
+        :raises IndexError:  if the length of the value is not 3.
+        :raises TypeError:  if the spacings are not positive numbers.
+        :raises ValueError:  if not enough information is available to calculate
+            spacings
+        """
+        if self._spacings is None:
+            if (self._lengths is not None) and (self._counts is not None):
+                spacings = []
+                for i in range(3):
+                    space = self._lengths[i] / (self._counts[i] - 1)
+                    spacings.append(space)
+                return spacings
+            else:
+                raise ValueError(
+                    f"Cannot calculate spacings because either lengths "
+                    f"{self._lengths} or counts {self._counts} is None."
+                )
+        return self._spacings
+
+    @spacings.setter
+    def spacings(self, value):
+        if len(value) != 3:
+            raise IndexError(f"Spacings are {len(value)} rather than 3.")
+        self._spacings = []
+        for i, elem in enumerate(value):
+            if not check.is_positive_definite(elem):
+                raise TypeError(
+                    f"Vector {value} does not contain positive numbers"
+                )
+            self._spacings.append(elem)
+
+    @property
+    def lengths(self) -> list:
+        """Length of the grid in a finite-difference Poisson-Boltzmann
+        calculation.
+
+        :returns: 3-vector of positive numbers indicating grid lengths in the
+          x-, y-, and z-directions in Å. The lengths may be different in each
+          direction.
+        :raises IndexError:  if the length of the value is not 3.
+        :raises TypeError:  if the lengths are not positive numbers.
+        :raises ValueError:  if not enough information is available to calculate
```

Comment:
Are TypeError and IndexError something that you plan to @TODO later?

Reply:
IndexError will get raised in line 296 if the vector length is less than 3.  Should I make this explicit?

TypeError will get raised in the setter (no way to document these separately in Sphinx) in line 313.

Reply:
Fair enough. I just didn't know if you wanted to try/except and exit out with a more "meaningful" error.